ykbh/src/lib.rs

```diff
@@ -117,70 +121,154 @@ impl SIRInterpreter {
-                    UnsignedInt::U8(v) => self.write(dest, [*v].as_ptr()),
-                    _ => todo!(),
-                },
-                ConstantInt::SignedInt(_si) => todo!(),
-            },
-            Constant::Bool(_b) => todo!(),
-            Constant::Tuple(t) => {
-                if SIR.ty(t).size() == 0 {
-                    // ZST: do nothing.
-                } else {
-                    todo!()
+    /// Returns a reference to the currently active frame.
+    fn frame(&self) -> &StackFrame {
+        self.frames.last().unwrap()
+    }
+
+    /// Returns a mutable reference to the currently active frame.
+    fn frame_mut(&mut self) -> &mut StackFrame {
+        self.frames.last_mut().unwrap()
+    }
+
+    /// Inserts a pointer to the trace inputs into `locals`.
+    pub fn set_trace_inputs(&mut self, tio: *mut u8) {
+        // FIXME Later this also sets other already initialised variables as well as the program
+        // counter of the interpreter.
+        let ptr = self.frame().local_ptr(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            // Write the pointer value of `tio` into locals.
+            std::ptr::write::<*mut u8>(ptr as *mut *mut u8, tio);
+        }
+    }
+
+    pub unsafe fn interpret(&mut self, body: ykpack::Body) {
+        // Ignore yktrace::trace_debug.
+        if body.flags.contains(BodyFlags::TRACE_DEBUG) {
+            return;
+        }
+
+        let mut bodies = vec![body];
+        let mut returns = Vec::new();
+        while let Some(body) = bodies.last() {
+            for stmt in body.blocks[self.bbidx].stmts.iter() {
+                match stmt {
+                    Statement::MkRef(dest, src) => self.mkref(dest, src),
+                    Statement::DynOffs { .. } => todo!(),
+                    Statement::Store(dest, src) => self.store(dest, src),
+                    Statement::BinaryOp { .. } => todo!(),
+                    Statement::Nop => {}
+                    Statement::Unimplemented(_) | Statement::Debug(_) => todo!(),
+                    Statement::Cast(..) => todo!(),
+                    Statement::Call(..) | Statement::StorageDead(_) => unreachable!(),
                 }
             }
-            _ => todo!(),
+
+            match &body.blocks[self.bbidx].term {
+                Terminator::Call {
+                    operand: op,
+                    args,
+                    destination: dest,
+                } => {
+                    let fname = if let CallOperand::Fn(sym) = op {
+                        sym
+                    } else {
+                        todo!("unknown call target");
+                    };
+
+                    // Initialise the new stack frame.
+                    let body = SIR.body(fname).unwrap();
+                    let mut frame = SIRInterpreter::allocate_locals(&body.local_decls);
+                    frame.copy_args(args, self.frame()); // Copy arguments to new frame.
```

Comment:
Redundant comment?

Reply:
Only if you think it's already clear from the code.

Reply:
`frame.copy_args(...)` is fairly clear to me :)

Reply:
I just wanted to be clear that we are copying from the old frame to the new frame. But will fix.

Reply:
Descriptive method names are always good. You could say `copy_args_in` or (more accurately?) `move_args_in` or similar?

Reply:
I agree that the comment is redundant.

Reply:
Fixed in d6d087a.